lib/Dialect/Verif/Transforms/PrepareVerifForExportVerilog.cpp

```diff
@@ -0,0 +1,91 @@
+//===- PrepareVerifForExportVerilog.cpp - Formal Preparations --------------*- C++ -*--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Prepare verif dialect for the verillog export.
+//
+//===----------------------------------------------------------------------===//
+#include "circt/Dialect/Verif/VerifOps.h"
+#include "circt/Dialect/Verif/VerifPasses.h"
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace circt;
+
+namespace circt {
+namespace verif {
+#define GEN_PASS_DEF_PREPAREVERIFFOREXPORTVERILOG
+#include "circt/Dialect/Verif/Passes.h.inc"
+} // namespace verif
+} // namespace circt
+
+
+using namespace mlir;
+using namespace verif;
+
+namespace {
+struct PrepareVerifForExportVerilogPass
+    : circt::verif::impl::PrepareVerifForExportVerilogBase<
+          PrepareVerifForExportVerilogPass> {
+  void runOnOperation() override;
+};
+
+struct FormalOpConversionPattern : public OpConversionPattern<verif::FormalOp> {
+  using OpConversionPattern<FormalOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(FormalOp op, OpAdaptor operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Create the ports for all the symbolic values
+    SmallVector<hw::PortInfo> ports;
+    for (auto sym : op.getBody().front().getOps<verif::SymbolicValueOp>()) {
```

Comment:
Is it guaranteed that verif::SymbolicValueOp appears in the top-level?

Reply:
It is because it has the `HasParent` trait.